recipes_source/zero_redundancy_optimizer.rst

```diff
@@ -0,0 +1,154 @@
+Shard Optimizer States with ZeroRedundancyOptimizer
+===================================================
+
+.. note:
+    `ZeroRedundancyOptimizer` is introduced in PyTorch 1.8 as a prototype
+    feature. It API is subject to change.
+
+In this recipe, you will learn:
+
+- The high-level idea of ``ZeroRedundancyOptimizer``.
+- How to use ``ZeroRedundancyOptimizer`` in distributed training and its impact.
+
+
+Requirements
+------------
+
+- PyTorch 1.8+
+- `Getting Started With Distributed Data Parallel <https://pytorch.org/tutorials/intermediate/ddp_tutorial.html>`_
+
+
+What is ``ZeroRedundancyOptimizer``?
+------------------------------------
+
+The idea of ``ZeroRedundancyOptimizer`` comes from the
+`DeepSpeed/ZeRO project <https://github.com/microsoft/DeepSpeed>`_ that shards
+optimizer states across distributed data-parallel processes to
+reduce per-process memory footprint. In the
+`Getting Started With Distributed Data Parallel <https://pytorch.org/tutorials/intermediate/ddp_tutorial.html>`_
+tutorial, we have shown how to use
+`DistributedDataParallel <https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html>`_
+(DDP) to train models. In that tutorial, each process keeps a dedicated replica
+of the optimizer. Since DDP has already synchronized gradients in the
+backward pass, all optimizer replicas will operate on the same parameter and
+gradient values in every iteration, and this is how DDP keeps model replicas in
+the same state. Oftentimes, optimizers also maintain local states. For example,
+the ``Adam`` optimizer uses per-parameter ``exp_avg`` and ``exp_avg_sq`` states. As a
+result, the ``Adam`` optimizer's memory consumption is at least twice the model
+size. Given this observation, we can reduce the optimizer memory footprint by
+sharding optimizer states across DDP processes. More specifically, instead of
+creating per-param states for all parameters, each optimizer instance in
+different DDP processes only keeps optimizer states for a shard of all model
+parameters. The optimizer ``step()`` function only updates the parameters in its
+shard and then broadcasts its updated parameters to all other peer DDP
+processes, so that all model replicas still land in the state across.
```

Comment:
>> so that all model replicas still land in the state across.
in general I think that the explanation is very good, but I don't get this sentence, looks like something is missing ?

Reply:
I should have removed "across". Thanks for spotting this!